sql/pipelines/src/main/scala/org/apache/spark/sql/pipelines/graph/TriggeredGraphExecution.scala

```diff
@@ -0,0 +1,469 @@
+      // Take flows that have terminated but have retry attempts left and flows that are queued, and
+      // filter the ones whose parents have all successfully completed, excluded, or idled because
+      // they are ONCE flows which already ran.
+      runnableFlows ++= (queuedForRetry ++ flowsWithState(StreamState.QUEUED)).filter { id =>
+        graphForExecution
+          .upstreamFlows(id)
+          .intersect(graphForExecution.materializedFlowIdentifiers)
+          .forall { id =>
+            pipelineState(id) == StreamState.SUCCESSFUL ||
+            pipelineState(id) == StreamState.EXCLUDED ||
+            pipelineState(id) == StreamState.IDLE
+          }
+      }
+
+      // collect flow that are ready to start
+      val flowsToStart = mutable.ArrayBuffer[ResolvedFlow]()
+      while (runnableFlows.nonEmpty && concurrencyLimit.tryAcquire()) {
+        val flowIdentifier = runnableFlows.head
+        runnableFlows.remove(flowIdentifier)
+        flowsToStart.append(graphForExecution.resolvedFlow(flowIdentifier))
+      }
+
+      val (batchFlowsToStart, otherFlowsToStart) = flowsToStart.partition { f =>
+        graphForExecution.resolvedFlow(f.identifier).isInstanceOf[CompleteFlow]
+      }
+
+      def startFlowWithPlanningMode(flow: ResolvedFlow, mode: String): Unit = {
+        val flowIdentifier = flow.identifier
+        logInfo(s"Starting flow ${flow.identifier} in $mode mode")
+        env.flowProgressEventLogger.recordPlanningForBatchFlow(flow)
+        try {
+          val flowStarted = startFlow(flow)
+          if (flowStarted.nonEmpty) {
+            pipelineState.put(flowIdentifier, StreamState.RUNNING)
+            logInfo(s"Flow $flowIdentifier started.")
+          } else {
+            if (flow.once) {
+              // ONCE flows are marked as IDLE in the event buffer for consistency with continuous
+              // execution where all unstarted flows are IDLE.
+              env.flowProgressEventLogger.recordIdle(flow)
+              pipelineState.put(flowIdentifier, StreamState.IDLE)
+              concurrencyLimit.release()
+            } else {
+              env.flowProgressEventLogger.recordSkipped(flow)
+              concurrencyLimit.release()
+              pipelineState.put(flowIdentifier, StreamState.SKIPPED)
+            }
+          }
+        } catch {
+          case NonFatal(ex) => recordFailed(flowIdentifier, ex)
+        }
+      }
+
+      // start non-batch flows serially because the configs will be attached to the pipeline's spark
+      // session (source dataframe's spark session)
+      otherFlowsToStart.foreach(startFlowWithPlanningMode(_, SERIAL_PLANNING))
+
+      // only start MV flows in parallel if enabled
+      startPlanning(batchFlowsToStart.toSeq) { (flow, mode) =>
+        startFlowWithPlanningMode(flow, mode)
+      }
+
+      try {
+        Thread.sleep(pipelineConf.streamStatePollingInterval * 1000)
```

Comment:
What's the reason that we sleep here?

Reply:
I think it's to avoid busy waiting, but let me double check with Ryan